src/confluent_kafka/aio/producer/_kafka_batch_executor.py

```diff
@@ -0,0 +1,109 @@
+            
+        Raises:
+            Exception: Any exception from the batch operation is propagated
+        """
+        def _produce_batch_and_poll():
+            """Helper function to run in thread pool
+            
+            This function encapsulates all the blocking Kafka operations:
+            - Call produce_batch with individual message callbacks
+            - Handle partial batch failures for messages that fail immediately
+            - Poll for delivery reports to trigger callbacks for successful messages
+            """
+            # Call produce_batch with individual callbacks (no batch callback)
+            # Convert tuple to list since produce_batch expects a list
+            messages_list = list(batch_messages) if isinstance(batch_messages, tuple) else batch_messages
+            self._producer.produce_batch(topic, messages_list)
+            
+            # Handle partial batch failures: Check for messages that failed during produce_batch
+            # These messages have their msgstates destroyed in Producer.c and won't get callbacks
+            # from librdkafka, so we need to manually invoke their callbacks
+            self._handle_partial_failures(messages_list)
+            
+            # Immediately poll to process delivery callbacks for successful messages
+            poll_result = self._producer.poll(0)
+            
+            return poll_result
+        
+        # Execute in thread pool to avoid blocking event loop
+        loop = asyncio.get_running_loop()
+        return await loop.run_in_executor(self._executor, _produce_batch_and_poll)
+    
+    def _handle_partial_failures(self, batch_messages):
+        """Handle messages that failed during produce_batch
+        
+        When produce_batch encounters messages that fail immediately (e.g., message too large,
+        invalid topic, etc.), librdkafka destroys their msgstates and won't call their callbacks.
+        We detect these failures by checking for '_error' in the message dict (set by Producer.c)
+        and manually invoke their callbacks.
+        
+        Args:
+            batch_messages: List of message dictionaries that were passed to produce_batch
+        """
+        for msg_dict in batch_messages:
+            if '_error' in msg_dict:
+                # This message failed during produce_batch - its callback won't be called by librdkafka
+                callback = msg_dict.get('callback')
+                if callback:
+                    # Extract the error from the message dict (set by Producer.c)
+                    error = msg_dict['_error']
+                    # Manually invoke the callback with the error
+                    # Note: msg is None since the message failed before being queued
+                    try:
+                        callback(error, None)
+                    except Exception as e:
+                        logger.error(f"Error in callback for failed message: {e}", exc_info=True)
```

Comment:
Repeat of don't swallow exceptions

Reply:
In this case you could ensure each callback is called in the partial list before raising the first exception hit... but that's a bit awkward too. I think it's probably better to just raise and tell callers to ensure exceptions aren't thrown by their code